### PR TITLE
Dummy inference engine #325

### PR DESCRIPTION
**Pull request by ji-cryptocafe**
Implemented **DummyInferenceEngine** to simulate inference without loading a model or running actual inference. The engine supports:

Static output mode, returning predefined values.
Random output mode, generating random outputs based on a specified shape.
Customizable latency using asyncio.sleep to simulate inference time with configurable mean and standard deviation.
All functionality is asynchronous to meet the requirements of non-blocking code.

**Testing**:

Added unit tests to verify functionality, including:
Static output mode.
Random output mode with shape and latency validation.
Latency checks to ensure it falls within the expected range.

### CODE REVIEW

**Comment by AlexCheema** [comment]
Looks good!

Can you add this as an option to the cli too? `--inference-engien dummy`

File: exo/inference/test_dummy_inference_engine.py [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+
+@pytest.mark.asyncio
+async def test_dummy_inference_engine_latency():
+    # Test that latency is within expected range
+    dummy_engine = DummyInferenceEngine(output_type="static", output_value=[1], latency_mean=0.1, latency_stddev=0.02)
+    start_time = asyncio.get_event_loop().time()
+    await dummy_engine.run_inference()
+    elapsed_time = asyncio.get_event_loop().time() - start_time
+    assert 0.06 <= elapsed_time <= 0.14, f"Expected latency to be around 0.1s, but got {elapsed_time}s."
```
Comment (AlexCheema) [inline exo/inference/test_dummy_inference_engine.py:30]: This is going to fail sometimes. Either make this so that probability of this failing is 1 in 1e18 (with multiple runs or just tuning stddev) or make stddev zero.

File: exo/inference/DummyInferenceEngine.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    latency = await dummy_engine.get_latency()
+    
+    assert isinstance(output, list), "Output should be a list."
+    assert isinstance(latency, float), "Latency should be a float."
+    
+    print(f"Simulated Inference Output: {output}")
+    print(f"Simulated Inference Latency: {latency}s")
+
+# Run the test
+# asyncio.run(test_dummy_engine())
```
Comment (AlexCheema) [inline exo/inference/DummyInferenceEngine.py:49]: Why is this here?

File: exo/inference/DummyInferenceEngine.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+            self.output_value = np.random.randn(*self.output_shape).tolist()  # Generate random output and store it
+            return self.output_value
+
+    async def get_latency(self):
+        # Simulate and return the latency
+        return max(0, random.normalvariate(self.latency_mean, self.latency_stddev))
+
+
+# Example usage and testing
+async def test_dummy_engine():
```
Comment (AlexCheema) [inline exo/inference/DummyInferenceEngine.py:35]: Can this go in the test?

File: exo/inference/test_dummy_inference_engine.py (hunk 1 above)
Comment (ji-cryptocafe) [reply exo/inference/test_dummy_inference_engine.py:30]: yep its quite sensitive. thought of let it run for a few 100k cycles to get a better idea of mean and stddev and also on different machines, but i guess your suggestion is the more pragmatic approach.

**Comment by ji-cryptocafe** [comment]
updated the code and added the cli option.

File: exo/inference/DummyInferenceEngine.py [hunk 4]
```diff
@@ -0,0 +1,32 @@
+import random  
+import asyncio
+import numpy as np  
+
+class DummyInferenceEngine:
```
Comment (AlexCheema) [inline exo/inference/DummyInferenceEngine.py:5]: This needs to implement the `InferenceEngine` interface

File: exo/main.py [hunk 5]
```diff
@@ -41,7 +41,7 @@
 parser.add_argument("--chatgpt-api-port", type=int, default=8000, help="ChatGPT API port")
 parser.add_argument("--chatgpt-api-response-timeout", type=int, default=90, help="ChatGPT API response timeout in seconds")
 parser.add_argument("--max-generate-tokens", type=int, default=10000, help="Max tokens to generate in each request")
-parser.add_argument("--inference-engine", type=str, default=None, help="Inference engine to use")
+parser.add_argument("--inference-engine", type=str, default=None, help="Inference engine to use e.g. 'mlx', 'tinygrad', 'dummy')")
```
Comment (AlexCheema) [inline exo/main.py:44]: This doesn't actually resolve the inference engine to `dummy` you need to change the code for that too. Please think through your code changes as right now this doesn't fit together at all. I'd like you to run this end-to-end with the `DummyInferenceEngine` before you submit it.

**Comment by ji-cryptocafe** [comment]
@AlexCheema clarification is needed:
Does the dummy inference only need to be run locally, without involving any other nodes in the network? 
In a full-fledged sharded version the requirement could be for the dummy inference engine to shard the pseudo work among the nodes communicating with each other while all of them running the dummy inference engine.

**Comment by AlexCheema** [comment]
> @AlexCheema clarification is needed: Does the dummy inference only need to be run locally, without involving any other nodes in the network? In a full-fledged sharded version the requirement could be for the dummy inference engine to shard the pseudo work among the nodes communicating with each other while all of them running the dummy inference engine.

Both need to be possible. This is outside of the scope of the `InferenceEngine` - the networking and orchestration between nodes happens outside of the `InferenceEngine`

**Comment by ji-cryptocafe** [comment]
ready for review

**Comment by AlexCheema** [comment]
Dummy implemented in a separate PR.
Please email alex@exolabs.net with your Ethereum address to claim $100 bounty.
